librz/analysis/arch/arm/arm_il32.c

```diff
@@ -0,0 +1,2695 @@
+	"q0", "q1", "q2", "q3", "q4", "q5", "q6", "q7", "q8", "q9", "q10", "q11", "q12", "q13", "q14", "q15",
+	NULL
+};
+
+/**
+ * Variable name for a register given by cs
+ */
+static const char *reg_var_name(arm_reg reg) {
+	switch (reg) {
+	case ARM_REG_LR: return "lr";
+	case ARM_REG_SP: return "sp";
+	case ARM_REG_Q0: return "q0";
+	case ARM_REG_Q1: return "q1";
+	case ARM_REG_Q2: return "q2";
+	case ARM_REG_Q3: return "q3";
+	case ARM_REG_Q4: return "q4";
+	case ARM_REG_Q5: return "q5";
+	case ARM_REG_Q6: return "q6";
+	case ARM_REG_Q7: return "q7";
+	case ARM_REG_Q8: return "q8";
+	case ARM_REG_Q9: return "q9";
+	case ARM_REG_Q10: return "q10";
+	case ARM_REG_Q11: return "q11";
+	case ARM_REG_Q12: return "q12";
+	case ARM_REG_Q13: return "q13";
+	case ARM_REG_Q14: return "q14";
+	case ARM_REG_Q15: return "q15";
+	case ARM_REG_R0: return "r0";
+	case ARM_REG_R1: return "r1";
+	case ARM_REG_R2: return "r2";
+	case ARM_REG_R3: return "r3";
+	case ARM_REG_R4: return "r4";
+	case ARM_REG_R5: return "r5";
+	case ARM_REG_R6: return "r6";
+	case ARM_REG_R7: return "r7";
+	case ARM_REG_R8: return "r8";
+	case ARM_REG_R9: return "r9";
+	case ARM_REG_R10: return "r10";
+	case ARM_REG_R11: return "r11";
+	case ARM_REG_R12: return "r12";
+	default: return NULL;
+	}
+}
+
+static ut32 reg_bits(arm_reg reg) {
+	if (reg >= ARM_REG_D0 && reg <= ARM_REG_D31) {
+		return 64;
+	}
+	if (reg >= ARM_REG_Q0 && reg <= ARM_REG_Q15) {
+		return 128;
+	}
+	return 32;
+}
+
+/**
+ * IL to read the given capstone reg
+ */
+static RzILOpBitVector *read_reg(ut64 pc, arm_reg reg) {
+	if (reg == ARM_REG_PC) {
+		return U32(pc);
+	}
+	if (reg >= ARM_REG_D0 && reg <= ARM_REG_D31) {
+		ut32 idx = reg - ARM_REG_D0;
+		RzILOpBitVector *var = VARG(reg_var_name(ARM_REG_Q0 + idx / 2));
+		return UNSIGNED(64, idx % 2 ? SHIFTR0(var, UN(7, 64)) : var);
+	}
+	if (reg >= ARM_REG_S0 && reg <= ARM_REG_S31) {
+		ut32 idx = reg - ARM_REG_S0;
+		RzILOpBitVector *var = VARG(reg_var_name(ARM_REG_Q0 + idx / 4));
+		return UNSIGNED(32, idx % 4 ? SHIFTR0(var, UN(7, 32 * (idx % 4))) : var);
+	}
+	const char *var = reg_var_name(reg);
+	return var ? VARG(var) : NULL;
+}
+
+#define PC(addr, is_thumb)      (addr + (is_thumb ? 4 : 8))
+#define PCALIGN(addr, is_thumb) (PC(addr, is_thumb) & ~3)
+#define REG_VAL(id)             read_reg(PC(insn->address, is_thumb), id)
+#define REG(n)                  REG_VAL(REGID(n))
+#define MEMBASE(x)              REG_VAL(insn->detail->arm.operands[x].mem.base)
+
+/**
+ * IL to write the given capstone reg
+ */
+static RzILOpEffect *write_reg(arm_reg reg, RZ_OWN RZ_NONNULL RzILOpBitVector *v) {
+	rz_return_val_if_fail(v, NULL);
+	if (reg >= ARM_REG_D0 && reg <= ARM_REG_D31) {
+		ut32 idx = reg - ARM_REG_D0;
+		arm_reg qreg = ARM_REG_Q0 + idx / 2;
+		ut64 mask[2] = { UT64_MAX, UT64_MAX };
+		mask[idx % 2] = 0;
+		RzILOpBitVector *masked = LOGAND(read_reg(0, qreg), rz_il_op_new_bitv(rz_bv_new_from_bytes_le((const ut8 *)mask, 0, 128)));
```

Comment:
What is 128 here?

Reply:
Size of a q reg, but I am rewriting this anyway.